app/models/voxelytics/LokiClient.scala

```diff
@@ -0,0 +1,173 @@
+      for {
+        _ <- akka.pattern.after(POLLING_INTERVAL, using = system.scheduler)(Future.successful(()))
+        _ <- bool2Fox(!until.isPast) ?~> s"Loki did not become ready within ${conf.startupTimeout}."
+        _ <- pollUntilServerStartedUp(until)
+      } yield ()
+
+    for {
+      isServerAvailableBox <- rpc(s"${conf.uri}/ready").request
+        .withMethod("GET")
+        .execute()
+        .flatMap(result =>
+          if (Status.isSuccessful(result.status)) {
+            Fox.successful(true)
+          } else if (result.status >= 500 && result.status < 600) {
+            logger.debug(s"Loki status: ${result.status}")
+            Fox.successful(false)
+          } else {
+            Fox.failure(s"Unexpected error code from Loki ${result.status}.")
+        })
+        .recoverWith({
+          case e: java.net.ConnectException =>
+            logger.debug(s"Loki connection exception: $e")
+            Fox.successful(false)
+          case e =>
+            logger.error(s"Unexpected error $e")
+            Fox.failure("Unexpected error while trying to connect to Loki.", Full(e))
+        })
+      isServerAvailable <- isServerAvailableBox.toFox
+      _ <- if (!isServerAvailable) {
+        waitAndRecurse(until)
+      } else {
+        logger.info("Loki is available.")
+        Fox.successful(())
+      }
+    } yield ()
+  }
+
+  def queryLogs(runName: String,
+                organizationId: ObjectId,
+                taskName: Option[String],
+                minLevel: VoxelyticsLogLevel = VoxelyticsLogLevel.INFO,
+                startTime: Instant,
+                endTime: Instant,
+                limit: Long,
+                backward: Boolean = true): Fox[JsValue] = {
+    val levels = VoxelyticsLogLevel.sortedValues.drop(VoxelyticsLogLevel.sortedValues.indexOf(minLevel))
+
+    val logQLFilter = List(
+      taskName.map(t => s"""vx_task_name="$t""""),
+      Some(s"""level=~"(${levels.mkString("|")})"""")
+    ).flatten.mkString(" | ")
+    val logQL =
+      s"""{vx_run_name="$runName",wk_org="${organizationId.id}",wk_url="${wkConf.Http.uri}"} | json vx_task_name,level | $logQLFilter"""
+
+    val queryString =
+      List("query" -> logQL,
+           "start" -> startTime.toString,
+           "end" -> endTime.toString,
+           "limit" -> limit.toString,
+           "direction" -> (if (backward) {
+                             "backward"
+                           } else {
+                             "forward"
+                           })).map(kv => s"${kv._1}=${java.net.URLEncoder.encode(kv._2, "UTF-8")}").mkString("&")
+    for {
+      _ <- serverStartupFuture
+      res <- rpc(s"${conf.uri}/loki/api/v1/query_range?$queryString").silent.getWithJsonResponse[JsValue]
+    } yield
+      JsArray(
+        (res \ "data" \ "result")
+          .as[List[JsValue]]
+          .flatMap(
+            stream =>
+              (stream \ "values")
+                .as[List[(String, String)]]
```

Comment:
Did you check what happens if the received json does not match the expected schema? Since this happens down here in the yield statement, it may become a bare exception that bubbles to the user. It might be worth it to move this up to the for part and wrap it with a tryo or one of the json parsing convenience functions that we already have (those may not suit this nested processing, though)